examples/TestEm3/gammas.cu

```diff
@@ -104,15 +109,57 @@ void TransportGammas(Track *gammas, const adept::MParray *active, Secondaries se
     // (Will be resampled in the next iteration.)
     currentTrack.numIALeft[winnerProcessIndex] = -1.0;
 
-    // Perform the discrete interaction.
-    RanluxppDoubleEngine rnge(&currentTrack.rngState);
+    g_nextInteractionForGamma[i] = winnerProcessIndex;
+    currentTrack.fPEmxSec = gammaTrack.GetPEmxSec();
+  }
+}
+
+
+template<int ProcessIndex>
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void ComputeGammaInteractions(Track *gammas, const adept::MParray *active, Secondaries secondaries,
+                              adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                              ScoringPerVolume *scoringPerVolume)
+{
+  constexpr int sharedSize = 12250;
+  __shared__ int candidates[sharedSize];
+  __shared__ unsigned int counter;
+  __shared__ unsigned int noopCounter;
+  counter = 0;
+  noopCounter = 0;
+
+  const int activeSize = active->size();
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const auto winnerProcess = g_nextInteractionForGamma[i];
+
+    if (winnerProcess == ProcessIndex) {
+      const auto destination = atomicInc(&counter, (unsigned int)-1);
+      candidates[destination % sharedSize] = i;
+    } else atomicInc(&noopCounter, (unsigned int)-1);
+  }
```

Comment:
This is a race condition, there needs to be a `__syncthreads()` in between. This is at least one of the reasons I was seeing non-reproducible results on my machine (when forcing `-batch 104`, the previous default)